tests/basic/tag.rs

```diff
@@ -1,70 +1,111 @@
+                assert_eq!(get_tag_for_current_task(), tag_on_entry);
+                let new_tag = f(tag_on_entry, i);
+                set_tag_for_current_task(new_tag);
+                assert_eq!(get_tag_for_current_task(), new_tag);
             })
         })
         .collect();
 
     block_on(join_all(threads));
+
+    assert_eq!(get_tag_for_current_task(), tag_on_entry);
 }
 
-fn spawn_some_threads_and_set_tag<F: (Fn(u64, u64) -> u64) + Send + Sync>(
-    tag_on_entry: u64,
+fn spawn_some_threads_and_set_tag<F: (Fn(Tag, u64) -> Tag) + Send + Sync>(
+    tag_on_entry: Tag,
     f: &'static F,
     num_threads: u64,
 ) {
     let threads: Vec<_> = (0..num_threads)
         .map(|i| {
             thread::spawn(move || {
-                assert!(get_tag() == tag_on_entry);
-                set_tag(f(tag_on_entry, i));
-                assert!(get_tag() == f(tag_on_entry, i));
+                assert_eq!(get_tag_for_current_task(), tag_on_entry);
+                let new_tag = f(tag_on_entry, i);
+                set_tag_for_current_task(new_tag);
+                assert_eq!(get_tag_for_current_task(), new_tag);
             })
         })
         .collect();
 
     threads.into_iter().for_each(|t| t.join().expect("Failed"));
 
-    assert!(get_tag() == tag_on_entry);
+    assert_eq!(get_tag_for_current_task(), tag_on_entry);
 }
 
-fn spawn_threads_which_spawn_more_threads(tag_on_entry: u64, num_threads: u64) {
-    let threads: Vec<_> = (0..num_threads)
+fn spawn_threads_which_spawn_more_threads(
+    tag_on_entry: Tag,
+    num_threads_first_block: u64,
+    num_threads_second_block: u64,
+) {
+    let mut threads: Vec<_> = (0..num_threads_first_block)
         .map(|i| {
             thread::spawn(move || {
-                assert!(get_tag() == tag_on_entry);
-                set_tag(i);
-                assert!(get_tag() == i);
-                spawn_some_threads_and_set_tag(i, &|_, _| 123, 13);
-                assert!(get_tag() == i);
-                spawn_some_threads_and_set_tag(i, &|_, x| x * 13, 7);
-                assert!(get_tag() == i);
-                spawn_some_threads_and_set_tag(i, &|p, x| (p << 4) + x, 19);
-                assert!(get_tag() == i);
-                spawn_some_futures_and_set_tag(i, &|p, x| (p << 4) & x, 17);
-                assert!(get_tag() == i);
+                assert_eq!(get_tag_for_current_task(), tag_on_entry);
+                let new_tag = i.into();
+                set_tag_for_current_task(new_tag);
+                assert_eq!(get_tag_for_current_task(), new_tag);
+                spawn_some_threads_and_set_tag(new_tag, &|_, _| 123.into(), 13);
+                assert_eq!(get_tag_for_current_task(), new_tag);
+                spawn_some_threads_and_set_tag(new_tag, &|_, x| (x * 13).into(), 7);
+                assert_eq!(get_tag_for_current_task(), new_tag);
+                spawn_some_threads_and_set_tag(new_tag, &|p, x| ((u64::from(p) << 4) + x).into(), 19);
```

Comment:
Feels like three calls to `spawn_some_threads_and_set_tag` are a bit overkill since they